File: 333.py
class Solution:
    def largestBSTSubtree(self, root: TreeNode) -> int:

        self.res = 0
        def isValid(root, t1, t2):
            if not root:
                return 0
            if root.val >= t2 or root.val <= t1:
                return float("-inf")
            return 1 + isValid(root.left, t1, root.val) + isValid(root.right, root.val, t2)

        def helper(root):
            if not root:
                return 
            self.res = max(self.res, isValid(root,float("-inf"), float("inf")) )
            helper(root.left)
            helper(root.right)
        helper(root)
        return self.res

File: 333.java
public class Solution {
    
    int max = 0;
    public int largestBSTSubtree(TreeNode root) {
        _largestBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
        return max;
    }
    
    // use null value as sign of broken BST condition
    private Integer _largestBST(TreeNode node, int minVal, int maxVal){
        if (node == null) return 0;
        
        Integer left = _largestBST(node.left, minVal, node.val);
        Integer right = _largestBST(node.right, node.val, maxVal);
        
        // independent research of subtrees has sense only if they aren't BST 
		// with current limitations on min and max
        if (left == null) _largestBST(node.left, Integer.MIN_VALUE, Integer.MAX_VALUE);
        if (right == null) _largestBST(node.right, Integer.MIN_VALUE, Integer.MAX_VALUE);
        
        if (left == null || right == null) return null;
        
        if (minVal < node.val && node.val < maxVal) {
            max = Math.max(max, 1 + left + right);
            return 1 + left + right;
        }
        return null;
    }
}